src/Kestrel.Core/Internal/Http2/Http2Connection.cs

```diff
@@ -270,13 +271,19 @@ public async Task ProcessRequestsAsync<TContext>(IHttpApplication<TContext> appl
                     // Ensure aborting each stream doesn't result in unnecessary WINDOW_UPDATE frames being sent.
                     _inputFlowControl.StopWindowUpdates();
 
-                    foreach (var stream in _streams.Values)
+                    var tasks = new List<Task>(_streams.Count);
+
+                    foreach (var stream in _streams)
                     {
-                        stream.Abort(new IOException(CoreStrings.Http2StreamAborted, connectionError));
+                        stream.Value.Abort(new IOException(CoreStrings.Http2StreamAborted, connectionError));
+                        tasks.Add(stream.Value.ProcessingTask);
                     }
 
                     await _streamsCompleted.Task;
 
+                    // REVIEW: Do we need a timeout?
+                    await Task.WhenAll(tasks);
```

Comment:
Yes, timeout. The app code could block these from finishing.

Any better name than tasks?

Reply:
This is re-implementing _streamsCompleted where the only difference is the ordering inside OnRequestProcessingEnded. Can you re-order that method to get the same effect? Or can you remove _streamsCompleted as it appears redundant?

Reply:
Sounds good